lib/commands/file-movement.js

```diff
@@ -0,0 +1,103 @@
+import _ from 'lodash';
+import log from '../logger';
+import path from 'path';
+import { fs, mkdirp, util, zip } from 'appium-support';
+
+// List of env variables, that can be expanded in path
+const KNOWN_ENV_VARS = [
+  'APPDATA', 'LOCALAPPDATA',
+  'PROGRAMFILES', 'PROGRAMFILES(X86)',
+  'PROGRAMDATA', 'ALLUSERSPROFILE',
+  'TEMP', 'TMP',
+  'HOMEPATH', 'USERPROFILE', 'PUBLIC'
+];
+const commands = {};
+
+commands.pushFile = async function pushFile (remotePath, base64Data) {
+  if (remotePath.endsWith(path.sep)) {
+    log.errorAndThrow(`It is expected that remote path points to a file and not to a folder. ` +
+      `'${remotePath}' is given instead`);
+  }
+
+  if (_.isArray(base64Data)) {
+    // some clients (ahem) java, send a byte array encoding utf8 characters
+    // instead of a string, which would be infinitely better!
+    base64Data = Buffer.from(base64Data).toString('utf8');
+  }
+
+  const fullPath = resolveToAbsolutePath(remotePath);
+  await mkdirp(path.dirname(fullPath));
+  const content = Buffer.from(base64Data, 'base64');
+  await fs.writeFile(fullPath, content);
+};
+
+commands.pullFile = async function pullFile (remotePath) {
+  if (remotePath.endsWith(path.sep)) {
+    log.errorAndThrow(`It is expected that remote path points to a file and not to a folder. ` +
+      `'${remotePath}' is given instead`);
+  }
+  const fullPath = resolveToAbsolutePath(remotePath);
+  return (await util.toInMemoryBase64(fullPath)).toString();
+};
+
+commands.pullFolder = async function pullFolder (remotePath) {
+  const fullPath = resolveToAbsolutePath(remotePath);
+  return (await zip.toInMemoryZip(fullPath, {
+    encodeToBase64: true,
+  })).toString();
+};
+
+/**
+ * @typedef {Object} DeleteFileOptions
+ * @property {string} remotePath - The path to a file
+ */
+
+/**
+ * Remove the file from the file system
+ *
+ * @param {DeleteFileOptions} opts
+ */
+commands.windowsDeleteFile = async function windowsDeleteFile (opts = {}) {
+  const { remotePath } = opts;
+  if (remotePath.endsWith(path.sep)) {
+    log.errorAndThrow(`It is expected that remote path points to a file and not to a folder. ` +
+      `'${remotePath}' is given instead`);
+  }
+  const fullPath = resolveToAbsolutePath(remotePath);
+  await fs.unlink(fullPath);
+};
+
+/**
+ * @typedef {Object} DeleteFolderOptions
+ * @property {string} remotePath - The path to a folder
+ */
+
+/**
+ * Remove the folder from the file system
+ *
+ * @param {DeleteFolderOptions} opts
+ */
+commands.windowsDeleteFolder = async function windowsDeleteFolder (opts = {}) {
+  const { remotePath } = opts;
```

Comment:
do we want to somehow handle the situation hen the remote folder does not exist?

Reply:
I looked at other drivers (ios, android) and did not see a general approach for appium drivers, how check input arguments. It is preferable to do input validation (which will decrease performance) or get an error from the called function? In this case `fs.unlink` throw exception that file does not exists.

Reply:
I would prefer an explicit file existence check. Also, we could decide to return either true or false instead of throwing an exception if the target file/folder does not exist

Reply:
I think the same response as the specification commands (`push_file`, etc.) would make it easier to understand and work with drivers. Because `response null` in this commands, I would prefer throwing exception.

Reply:
Added file/folder exists check.